lib/kubernetes-deploy/discoverable_resource.rb

```diff
@@ -0,0 +1,232 @@
+      @child_classes = Set.new
+      resources = client.get_custom_resource_definitions
+      resources.each do |res|
+        kind = res.spec.names.kind
+        # Remove and redefine the class if it already exists so we can be up to date.
+        if DiscoverableResource.const_defined?(kind)
+          klass = DiscoverableResource.const_get(kind)
+          DiscoverableResource.send(:remove_const, kind)
+          child_classes.delete(klass)
+        end
+        generate_resource(group: res.spec.group,
+                           version: res.spec.version,
+                           kind: kind,
+                           annotations: res.metadata.annotations)
+        logger.info("  - #{res.spec.group}/#{res.spec.version}/#{kind}")
+      end
+    end
+
+    def self.generate_resource(group:, version:, kind:, annotations:)
+      deploy_metadata = annotations[DEPLOY_METADATA_ANNOTATION] || '{}'
+      metadata = JSON.parse(deploy_metadata)
+      unless metadata.is_a?(Hash)
+        raise FatalDeploymentError, "Invalid metadata for #{kind} #{metadata.inspect}"
+      end
+
+      predeploy_dependencies = [] << metadata['predeploy-dependencies']
+      predeploy_dependencies.flatten!
+      predeploy_dependencies.compact!
+
+      prunable = parse_bool(metadata['prunable'])
+      predeploy = parse_bool(metadata['predeploy'])
+      status_query = metadata['status-query']
+      success_status = metadata['status-success']
+
+      resource_template = ERB.new <<-CLASS
+        class #{kind.capitalize} < DiscoverableResource
+          GROUP = '#{group}'
+          VERSION = '#{version}'
+          PREDEPLOY = #{predeploy}
+          PRUNABLE = #{prunable}
+
+          <% unless predeploy_dependencies.empty? %>
+          PREDEPLOY_DEPENDENCIES = #{predeploy_dependencies}
+          <% end %>
+
+          <% if status_query && success_status %>
+          def deploy_succeeded?
+            getter = "get_#{kind.downcase}"
+            @client ||= DiscoverableResource.kubeclient(context: @context, resource_class: self.class)
+            raw_json = @client.send(getter, @name, @namespace, as: :raw)
+            result = query_json(query: STATUS_QUERY, json: raw_json)
+            result == SUCCESS_STATUS
+          end
+          <% end %>
+
+          self
+        end
+      CLASS
+
+      rendered_template = resource_template.result(binding)
+      klass = class_eval(rendered_template)
+      klass.const_set(:STATUS_QUERY, status_query)
+      klass.const_set(:SUCCESS_STATUS, success_status)
+      klass
+    rescue JSON::ParserError => e
+      raise FatalDeploymentError, "Invalid metadata for #{group}/#{version}/#{kind}: #{e}"
+    end
+
+    def self.parse_bool(value)
+      return true if TRUE_VALUES.include?(value)
+      false
+    end
+
+    def query_json(query:, json:)
+      safe_query = Shellwords.escape(query)
+      out, err, status = Open3.capture3("jq -r #{safe_query}", stdin_data: json)
```

Comment:
This approach is gross and has poor performance, the rationale:

- I initially used the `jsonpath` gem, but that query language is fairly obscure. `jq` is much more popular.
- I tried using the `ruby-jq` gem but it has two issues:
  - It requires a special shared library version of `jq` to be installed (this was doable, but annoying)
  - The `jq` parser prints errors to `stderr`, so even if you use the gem your program will print out error output from deep within the `jq` library itself. This was a show stopper. It might be possible to disable the output, but I did not see how with configure options and I do not want to have to maintain a fork of `jq`